javalib/src/main/scala/java/util/HashMap.scala

```diff
@@ -12,153 +12,502 @@
+    threshold = newThreshold(newlen)
+    if (contentSize == 0) {
+      table = new Array(newlen)
+    } else {
+      val newTable = new Array[Node[K, V]](newlen)
+      table = newTable
+
+      /* Split the nodes of each bucket from the old table into the "low" and
+       * "high" indices of the new table. Since the new table contains exactly
+       * twice as many buckets as the old table, every index `i` from the old
+       * table is split into indices `i` and `oldlen + i` in the new table.
+       */
+      var i = 0
+      while (i < oldlen) {
+        var lastLow: Node[K, V] = null
+        var lastHigh: Node[K, V] = null
+        var node = oldTable(i)
+        while (node ne null) {
+          if ((node.hash & oldlen) == 0) {
+            // go to low
+            node.previous = lastLow
+            if (lastLow eq null)
+              newTable(i) = node
+            else
+              lastLow.next = node
+            lastLow = node
+          } else {
+            // go to high
+            node.previous = lastHigh
+            if (lastHigh eq null)
+              newTable(oldlen + i) = node
+            else
+              lastHigh.next = node
+            lastHigh = node
+          }
+          node = node.next
+        }
+        if (lastLow ne null)
+          lastLow.next = null
+        if (lastHigh ne null)
+          lastHigh.next = null
+        i += 1
       }
     }
   }
 
-  private class ValuesView extends AbstractMapView[V] {
-    override def size(): Int =
-      inner.size
+  private[this] def tableSizeFor(capacity: Int): Int =
+    Math.min(Integer.highestOneBit(Math.max(capacity - 1, 4)) * 2, 1 << 30)
+
+  private[this] def newThreshold(size: Int) =
+    (size.toDouble * loadFactor).toInt
+
+  // Iterators
+
+  private[util] def nodeIterator(): ju.Iterator[Node[K, V]] =
+    new NodeIterator
+
+  private[util] def keyIterator(): ju.Iterator[K] =
+    new KeyIterator
+
+  private[util] def valueIterator(): ju.Iterator[V] =
+    new ValueIterator
+
+  // The cast works around the lack of definition-site variance
+  private[util] final def entrySetIterator(): ju.Iterator[Map.Entry[K, V]] =
+    nodeIterator().asInstanceOf[ju.Iterator[Map.Entry[K, V]]]
+
+  private final class NodeIterator extends AbstractHashMapIterator[Node[K, V]] {
+    protected[this] def extract(node: Node[K, V]): Node[K, V] = node
+  }
+
+  private final class KeyIterator extends AbstractHashMapIterator[K] {
+    protected[this] def extract(node: Node[K, V]): K = node.key
+  }
 
-    override def iterator(): Iterator[V] = {
-      new AbstractMapViewIterator[V] {
-        protected def getNextForm(key: Box[K]): V = inner(key)
+  private final class ValueIterator extends AbstractHashMapIterator[V] {
+    protected[this] def extract(node: Node[K, V]): V = node.value
+  }
+
+  private abstract class AbstractHashMapIterator[A] extends ju.Iterator[A] {
+    private[this] val len = table.length
+    private[this] var nextIdx: Int = _
```

Comment:
It's for performance (to avoid a double-assignment). But I'll add a comment that we do rely on the 0. (and same with the `null`s for `nextNode` and `lastNode`.